### ISSUE TITLE
Proxy wrapper should be removed when passing object back to napari functions

### ISSUE THREAD

**Issue by Czaki** [napari/napari#3782]
## 🐛 Bug

There is a possibility to generate Proxy future warnings using only public API, because the proxy is not removed when passing object bac to napari function (which uses private API). 

## To Reproduce

The sample scenario is to remove one layer to add it later. 

1.
```python

import numpy as np
from napari_plugin_engine import napari_hook_implementation

from napari import Viewer, run, plugins
from qtpy.QtWidgets import QWidget, QPushButton, QHBoxLayout




class RemoveAdd(QWidget):
    def __init__(self, viewer: Viewer):
        super().__init__()
        self.viewer = viewer
        self.layer = None
        self.btn = QPushButton("Remove/add", self)
        self.btn.clicked.connect(self._click)
        layout = QHBoxLayout(self.btn)
        self.setLayout(layout)

    def _click(self):
        if self.layer is None:
            self.layer = self.viewer.layers[-1]
            del self.viewer.layers[-1]
        else:
            # self.viewer.add_layer(self.layer.__wrapped__)  # This line works correctly
            self.viewer.add_layer(self.layer)
            self.layer = None

@napari_hook_implementation
def napari_experimental_provide_dock_widget():
    return RemoveAdd

plugin = plugins.plugin_manager.register(globals(), name="Test Widget")


viewer = Viewer()
viewer.add_image(np.random.random((20, 20)))
run()
```
2. open Test Widget RemoveAdd plugin. 
3. click "Remove/add" button twice.

## Expected behavior

No warnings

## Environment

This commit [`d3d090f` (#3777)](https://github.com/napari/napari/pull/3777/commits/d3d090fca3ffe8688602fa2dfb177d6da8f27521)
Running on last main cause exceptions connected with wrapping non-napari objects. 

## Additional context

Generated warnings:
```
/home/czaki/Projekty/napari/napari/_vispy/layers/image.py:84: FutureWarning: Private attribute access ('Image._ndisplay') in this context (e.g. inside a plugin widget or dock widget) is deprecated and will be unavailable in version 0.4.14!
/home/czaki/Projekty/napari/napari/_vispy/layers/image.py:87: FutureWarning: Private attribute access ('Image._ndisplay') in this context (e.g. inside a plugin widget or dock widget) is deprecated and will be unavailable in version 0.4.14!
/home/czaki/Projekty/napari/napari/_vispy/layers/image.py:89: FutureWarning: Private attribute access ('Image._empty') in this context (e.g. inside a plugin widget or dock widget) is deprecated and will be unavailable in version 0.4.14!
/home/czaki/Projekty/napari/napari/_vispy/layers/base.py:126: FutureWarning: Private attribute access ('Image._transforms') in this context (e.g. inside a plugin widget or dock widget) is deprecated and will be unavailable in version 0.4.14!
/home/czaki/Projekty/napari/napari/_vispy/layers/base.py:127: FutureWarning: Private attribute access ('Image._dims_displayed') in this context (e.g. inside a plugin widget or dock widget) is deprecated and will be unavailable in version 0.4.14!
/home/czaki/Projekty/napari/napari/_vispy/layers/base.py:140: FutureWarning: Private attribute access ('Image._ndisplay') in this context (e.g. inside a plugin widget or dock widget) is deprecated and will be unavailable in version 0.4.14!
/home/czaki/Projekty/napari/napari/_vispy/layers/base.py:145: FutureWarning: Private attribute access ('Image._data_to_world') in this context (e.g. inside a plugin widget or dock widget) is deprecated and will be unavailable in version 0.4.14!
/home/czaki/Projekty/napari/napari/_vispy/layers/base.py:146: FutureWarning: Private attribute access ('Image._dims_displayed') in this context (e.g. inside a plugin widget or dock widget) is deprecated and will be unavailable in version 0.4.14!
/home/czaki/Projekty/napari/napari/_vispy/layers/image.py:107: FutureWarning: Private attribute access ('Image._data_view') in this context (e.g. inside a plugin widget or dock widget) is deprecated and will be unavailable in version 0.4.14!
/home/czaki/Projekty/napari/napari/_vispy/layers/image.py:114: FutureWarning: Private attribute access ('Image._ndisplay') in this context (e.g. inside a plugin widget or dock widget) is deprecated and will be unavailable in version 0.4.14!
/home/czaki/Projekty/napari/napari/_vispy/layers/image.py:118: FutureWarning: Private attribute access ('Image._ndisplay') in this context (e.g. inside a plugin widget or dock widget) is deprecated and will be unavailable in version 0.4.14!
/home/czaki/Projekty/napari/napari/_vispy/layers/image.py:127: FutureWarning: Private attribute access ('Image._ndisplay') in this context (e.g. inside a plugin widget or dock widget) is deprecated and will be unavailable in version 0.4.14!
/home/czaki/Projekty/napari/napari/_vispy/layers/image.py:128: FutureWarning: Private attribute access ('Image._ndisplay') in this context (e.g. inside a plugin widget or dock widget) is deprecated and will be unavailable in version 0.4.14!
/home/czaki/Projekty/napari/napari/_vispy/layers/image.py:133: FutureWarning: Private attribute access ('Image._empty') in this context (e.g. inside a plugin widget or dock widget) is deprecated and will be unavailable in version 0.4.14!
/home/czaki/Projekty/napari/napari/_qt/layer_controls/qt_image_controls.py:52: FutureWarning: Private attribute access ('EmitterGroup._ndisplay') in this context (e.g. inside a plugin widget or dock widget) is deprecated and will be unavailable in version 0.4.14!
/home/czaki/Projekty/napari/napari/_qt/layer_controls/qt_image_controls.py:244: FutureWarning: Private attribute access ('Image._ndisplay') in this context (e.g. inside a plugin widget or dock widget) is deprecated and will be unavailable in version 0.4.14!
/home/czaki/Projekty/napari/napari/_qt/layer_controls/qt_image_controls.py:256: FutureWarning: Private attribute access ('Image._ndisplay') in this context (e.g. inside a plugin widget or dock widget) is deprecated and will be unavailable in version 0.4.14!
/home/czaki/Projekty/napari/napari/components/viewer_model.py:317: FutureWarning: Private attribute access ('Image._slice_dims') in this context (e.g. inside a plugin widget or dock widget) is deprecated and will be unavailable in version 0.4.14!
/home/czaki/Projekty/napari/napari/components/layerlist.py:53: FutureWarning: Private attribute access ('Image._on_selection') in this context (e.g. inside a plugin widget or dock widget) is deprecated and will be unavailable in version 0.4.14!
/home/czaki/Projekty/napari/napari/utils/context/_layerlist_context.py:47: FutureWarning: Private attribute access ('Image._type_string') in this context (e.g. inside a plugin widget or dock widget) is deprecated and will be unavailable in version 0.4.14!
/home/czaki/Projekty/napari/napari/utils/context/_layerlist_context.py:39: FutureWarning: Private attribute access ('Image._type_string') in this context (e.g. inside a plugin widget or dock widget) is deprecated and will be unavailable in version 0.4.14!
/home/czaki/Projekty/napari/napari/utils/context/_layerlist_context.py:43: FutureWarning: Private attribute access ('Image._type_string') in this context (e.g. inside a plugin widget or dock widget) is deprecated and will be unavailable in version 0.4.14!
/home/czaki/Projekty/napari/napari/_qt/qt_viewer.py:960: FutureWarning: Private attribute access ('Image._displayed_axes') in this context (e.g. inside a plugin widget or dock widget) is deprecated and will be unavailable in version 0.4.14!
/home/czaki/Projekty/napari/napari/_qt/qt_viewer.py:965: FutureWarning: Private attribute access ('Image._update_draw') in this context (e.g. inside a plugin widget or dock widget) is deprecated and will be unavailable in version 0.4.14!
/home/czaki/Projekty/napari/napari/_qt/containers/_layer_delegate.py:99: FutureWarning: Private attribute access ('Image._type_string') in this context (e.g. inside a plugin widget or dock widget) is deprecated and will be unavailable in version 0.4.14!
```

**Comment by tlambert03** [comment]
thanks!  agreed, that shouldn't warn

**Comment by tlambert03** [comment]
this warning is also fixed by #3777 ... I like the idea of removing the wrapper as well, but it's harder to think about how to do that (without adding a lot of instance checks everywhere).  need to ponder

**Comment by tlambert03** [comment]
(this is actually a variant of #3775)

**Comment by Czaki** [comment]
> this warning is also fixed by #3777 ... I like the idea of removing the wrapper as well, but it's harder to think about how to do that (without adding a lot of instance checks everywhere). need to ponder

as I commented in #3777. If you check that calling place is inside napari itself, you could return object, not proxy of object. Then first object still will be wrapped, but sub objects will not be wrapped.
